File: Binary-tree-advance-DS/binaryTree.js
class Node {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

// Minimal BFS-only Binary Tree
class BinaryTree {
  constructor() {
    this.root = null;
  }

  // BFS Insert
  insert(value) {
    const newNode = new Node(value);
    if (!this.root) {
      this.root = newNode;
      return;
    }
    const queue = [this.root];

    while (queue.length) {
      const current = queue.shift();
      if (!current.left) {
        current.left = newNode;
        return;
      }
      if (!current.right) {
        current.right = newNode;
        return;
      }
      queue.push(current.left, current.right);
    }
  }

  // BFS Traversal
  traverse() {
    if (!this.root) return [];

    const result = [];
    const queue = [this.root];

    while (queue.length) {
      const current = queue.shift();
      result.push(current.value);

      if (current.left) queue.push(current.left);
      if (current.right) queue.push(current.right);
    }

    return result;
  }

  // NEW: Depth First Search Traversals

  // 1. DFS Pre-order Traversal: Root → Left → Right
  preOrder(node = this.root, result = []) {
    if (node) {
      result.push(node.value);
      this.preOrder(node.left, result);
      this.preOrder(node.right, result);
    }
    return result;
  }

  // 2. DFS In-order: Left → Root → Right
  inOrder(node = this.root, result = []) {
    if (node) {
      this.inOrder(node.left, result);
      result.push(node.value);
      this.inOrder(node.right, result);
    }
    return result;
  }

  // 3. DFS Post-order: Left → Right → Root
  postOrder(node = this.root, result = []) {
    if (node) {
      this.postOrder(node.left, result);
      this.postOrder(node.right, result);
      result.push(node.value);
    }
    return result;
  }

  // BFS Search (Level-order search)
  searchBFS(value) {
    if (!this.root) return false;

    const queue = [this.root];

    while (queue.length) {
      const current = queue.shift();
      if (current.value === value) return true;

      if (current.left) queue.push(current.left);
      if (current.right) queue.push(current.right);
    }
    return false;
  }

  // DFS Search (using recursion - pre order )
  searchDFS(value, node = this.root) {
    if (!node) return false;

    //check current node
    if (node.value === value) return true;

    // Recursively search left and right subtrees
    return (
      this.searchDFS(value, node.left) || this.searchDFS(value, node.right)
    );
  }

  // Find Node (returns the node instead of boolean)
  findNode(value, node = this.root) {
    if (!node) return null;

    if (node.value === value) return node;

    const leftResult = this.findNode(value, node.left);
    if (leftResult) return leftResult;

    return this.findNode(value, node.right);
  }
}

// Quick Demo
const binaryTree = new BinaryTree();
[10, 20, 30, 40, 50].forEach((val) => binaryTree.insert(val));

console.log("Tree Structure:");
console.log("      10");
console.log("     /  \\");
console.log("    20   30");
console.log("   / \\");
console.log("  40  50");

console.log("BFS Traversal:", binaryTree.traverse());
// Output: [10, 20, 30, 40, 50]
console.log("DFS Pre-order Traversal:", binaryTree.preOrder());
// Output: [10, 20, 40, 50, 30]
console.log("DFS In-order Traversal:", binaryTree.inOrder());
// Output: [40, 20, 50, 10, 30]
console.log("DFS Post-order Traversal:", binaryTree.postOrder());
// Output: [40, 50, 20, 30, 10]
// Test searches
console.log("=== SEARCH OPERATIONS ===");
console.log("BFS Search for 50:", binaryTree.searchBFS(50)); // true
console.log("BFS Search for 99:", binaryTree.searchBFS(99)); // false

console.log("DFS Search for 40:", binaryTree.searchDFS(40)); // true
console.log("DFS Search for 25:", binaryTree.searchDFS(25)); // false

console.log("Find Node 50:", binaryTree.findNode(50)?.value); // 50
console.log("Find Node 80:", binaryTree.findNode(80)); // null

/* OUTPUT of Binary Tree Data Structure and Traversals: 
Tree Structure:
      10
     /  \
    20   30
   / \
  40  50
BFS Traversal: [ 10, 20, 30, 40, 50 ]
DFS Pre-order Traversal: [ 10, 20, 40, 50, 30 ]
DFS In-order Traversal: [ 40, 20, 50, 10, 30 ]
DFS Post-order Traversal: [ 40, 50, 20, 30, 10 ]
=== SEARCH OPERATIONS ===
BFS Search for 50: true
BFS Search for 99: false
DFS Search for 40: true
DFS Search for 25: false
*/

File: Binary-tree-advance-DS/binary_tree.rb
class Node
  attr_accessor :val, :left, :right
  
  def initialize(val)
    @val = val
    @left = nil
    @right = nil
  end
end

class BinaryTree
  def initialize
    @root = nil
  end
  
  def insert(val)
    node = Node.new(val)
    return @root = node if @root.nil?
    
    queue = [@root]
    until queue.empty?
      curr = queue.shift
      return curr.left = node if curr.left.nil?
      return curr.right = node if curr.right.nil?
      queue << curr.left << curr.right
    end
  end
  
  # BFS Traversal
  def bfs
    return [] if @root.nil?
    result, queue = [], [@root]
    until queue.empty?
      curr = queue.shift
      result << curr.val
      queue << curr.left if curr.left
      queue << curr.right if curr.right
    end
    result
  end

  # DFS Traversals

   # 1. DFS Pre-order: Root → Left → Right
   def pre_order(node = @root, result = [])
   if node 
    result << node.val
    pre_order(node.left, result)
    pre_order(node.right, result)
    
   end
   result
   end

    # 2. DFS In-order: Left → Root → Right
    def in_order(node = @root, result = [])
    if node
    in_order(node.left, result)
    result << node.val
    in_order(node.right, result)
    end
    result
    end

    # 3. DFS Post-order: Left → Right → Root
    def post_order(node = @root, result = [])
    if node 
    post_order(node.left, result)
    post_order(node.right, result)
    result << node.val
    end
    result
    end

     # SEARCH OPERATIONS

     # 1. BFS Search (Level-order search)
     def search_bfs(value)
      return false if @root.nil?

      queue = [@root]

      until queue.empty?
        current = queue.shift

        # Check if current node matches
        return true if current.val == value

        # Add children to queue
        queue << current.left if current.left
        queue << current.right if current.right
      end
      false
    end

     # 2. DFS Search (using recursion - pre order)
     def search_dfs(value, node = @root)
    return false if node.nil?
    
    # Check current node
    return true if node.val == value
    
    # Recursively search left and right subtrees
    search_dfs(value, node.left) || search_dfs(value, node.right)
  end

     # 3. Find Node (returns the node instead of boolean)
 def find_node(value, node = @root)
    return nil if node.nil?
    
    return node if node.val == value
    
    # Search left subtree first
    left_result = find_node(value, node.left)
    return left_result if left_result
    
    # Then search right subtree
    find_node(value, node.right)
  end

  # 4. Search with Path (returns the path to the node)
  def search_with_path(value, node = @root, path = [])
    return nil if node.nil?
    
    # Add current node to path
    path = path + [node.val]
    
    return path if node.val == value
    
    # Search left subtree
    left_path = search_with_path(value, node.left, path)
    return left_path if left_path
    
    # Search right subtree  
    right_path = search_with_path(value, node.right, path)
    return right_path if right_path
    
    nil # Not found in either subtree
  end

end

# Demo
tree = BinaryTree.new
[10, 20, 30, 40, 50].each { |x| tree.insert(x) }

puts "Tree Structure:"
puts "      10"
puts "     /  \\"
puts "    20   30"
puts "   / \\"
puts "  40  50"
puts

puts "BFS Traversal: #{tree.bfs.inspect}"
puts "DFS Pre-order: #{tree.pre_order.inspect}"
puts "DFS In-order: #{tree.in_order.inspect}"
puts "DFS Post-order: #{tree.post_order.inspect}"

# Test searches
puts "\n=== SEARCH OPERATIONS ==="
puts "BFS Search for 50: #{tree.search_bfs(50)}" # true
puts "BFS Search for 99: #{tree.search_bfs(99)}" # false

puts "DFS Search for 40: #{tree.search_dfs(40)}" # true  
puts "DFS Search for 25: #{tree.search_dfs(25)}" # false

puts "Find Node 50: #{tree.find_node(50)&.val}" # 50
puts "Find Node 80: #{tree.find_node(80)}" # nil

puts "Path to 50: #{tree.search_with_path(50)}" # [10, 20, 50]
puts "Path to 30: #{tree.search_with_path(30)}" # [10, 30]
puts "Path to 25: #{tree.search_with_path(25)}" # nil

=begin
Tree Structure:
      10
     /  \
    20   30
   / \
  40  50

BFS Traversal: [10, 20, 30, 40, 50]
DFS Pre-order: [10, 20, 40, 50, 30]
DFS In-order: [40, 20, 50, 10, 30]
DFS Post-order: [40, 50, 20, 30, 10]
=== SEARCH OPERATIONS ===
BFS Search for 50: true
BFS Search for 99: false
DFS Search for 40: true
DFS Search for 25: false
Find Node 50: 50
Find Node 80:
Path to 50: [10, 20, 50]
Path to 30: [10, 30]
Path to 25:
=end

